doc/man7/ossl-guide-quic-client-non-block.pod

```diff
@@ -0,0 +1,453 @@
+
+        /* Get hold of the underlying file descriptor for the socket */
+        sock = SSL_get_fd(ssl);
+
+        FD_ZERO(&wfds);
+        FD_ZERO(&rfds);
+
+        /*
+         * Find out if we would like to write to the socket, or read from it (or
+         * both)
+         */
+        if (SSL_net_write_desired(ssl))
+            FD_SET(sock, &wfds);
+        if (SSL_net_read_desired(ssl))
+            FD_SET(sock, &rfds);
+        width = sock + 1;
+
+        /*
+         * Find out when OpenSSL would next like to be called, regardless of
+         * whether the state of the underlying socket has changed or not.
+         */
+        if (SSL_get_event_timeout(ssl, &tv, &isinfinite) && !isinfinite)
+            tvp = &tv;
+
+        /*
+         * Wait until the socket is writeable or readable. We use select here
+         * for the sake of simplicity and portability, but you could equally use
+         * poll/epoll or similar functions
+         *
+         * NOTE: For the purposes of this demonstration code this effectively
+         * makes this demo block until it has something more useful to do. In a
+         * real application you probably want to go and do other work here (e.g.
+         * update a GUI, or service other connections).
+         *
+         * Let's say for example that you want to update the progress counter on
+         * a GUI every 100ms. One way to do that would be to use the timeout in
+         * the last parameter to "select" below. If the tvp value is greater
+         * than 100ms then use 100ms instead. Then, when select returns, you
+         * check if it did so because of activity on the file descriptors or
+         * because of the timeout. If the 100ms GUI timeout has expired but the
+         * tvp timeout has not then go and update the GUI and then restart the
+         * "select" (with updated timeouts).
+         */
+
+        select(width, &rfds, &wfds, NULL, tvp);
+}
+
+If you are familiar with how to write nonblocking applications in OpenSSL for
+TLS (see L<ossl-guide-tls-client-non-block(7)>) then you should note that there
+is an important difference here between the way a QUIC application and a TLS
+application works. With a TLS application if we try to read or write something
+to the B<SSL> object and we get a "retry" response (B<SSL_ERROR_WANT_READ> or
+B<SSL_ERROR_WANT_WRITE>) then we can assume that is because OpenSSL attempted to
+read or write to the underlying socket and the socket signalled the "retry".
+With QUIC that is not the case. OpenSSL may signal retry as a result of an
+L<SSL_read_ex(3)> or L<SSL_write_ex(3)> (or similar) call which indicates the
+state of the stream. This is entirely independent of whether the underlying
+socket needs to retry or not.
+
+To determine whether OpenSSL currently wants to read or write to the underlying
+socket for a QUIC application we must call the L<SSL_net_read_desired(3)> and
+L<SSL_net_write_desired(3)> functions.
```

Comment:
Could we trivially make SSL_net_read_desired() and SSL_net_write_desired() work with TLS/DTLS connections? I.e. just return the values based on what SSL_want_read() or SSL_want_write() returned?

Reply:
We could.

Reply:
https://github.com/openssl/openssl/pull/21979